File: frontend/src/lib/queries/pipeline.ts
'use client'

import { useMutation, useQuery } from '@tanstack/react-query'
import { useApi } from '@/lib/api/use-api'
import type { TablePreview } from '@/lib/queries/ingestion'

export type RelationshipCandidate = {
  table_a: string
  column_a: string
  table_b: string
  column_b: string
  confidence: number
  evidence: Record<string, number>
  source: string
  reasoning: string
  validation: {
    pk_uniqueness: number
    fk_coverage: number
    orphan_ratio: number
    dtype_compatible: boolean
  }
}

export type SchemaDiscoveryResult = {
  tables: { name: string; row_count: number; column_count: number; columns: string[] }[]
  auto: RelationshipCandidate[]
  review: RelationshipCandidate[]
  manual: RelationshipCandidate[]
}

export type RelationshipDecision = {
  table_a: string
  column_a: string
  table_b: string
  column_b: string
  confidence: number
  evidence: Record<string, number>
  status: 'approved' | 'rejected'
}

export type RelationshipOut = RelationshipDecision & { approved_by: string }

export function decisionKey(c: { table_a: string; column_a: string; table_b: string; column_b: string }) {
  return `${c.table_a}.${c.column_a}->${c.table_b}.${c.column_b}`
}

export type CleaningPlanStep = {
  id: string
  description: string
  /** 'detector' = derived deterministically from measured defects, 'generated' = LLM-refined,
   *  'manual' = typed by the user, 'fallback' = produced after an LLM failure. */
  source: 'generated' | 'manual' | 'detector' | 'fallback'
  /** Name of the deterministic operator this step runs. Empty means a free-text
   *  instruction, which is executed by generated code instead. */
  op?: string
  columns?: string[]
  params?: Record<string, any>
}

export type PlanResult = {
  table_name: string
  profile: Record<string, any>
  cleaning_plan: CleaningPlanStep[]
}

export type CleanResult = {
  table_name: string
  generated_code: string
  success: boolean
  error: string | null
  validation_report: Record<string, any>
  transformation_log: string[]
  retry_count: number
  preview: TablePreview | null
  /** Measured per-step audit trail: counts computed by diffing the table
   *  before and after each step, not reported by the code that changed it. */
  ledger: Record<string, any> | null
}

export type TableCleaningStatus = {
  table_name: string
  success: boolean
  error: string | null
  validation_report: Record<string, any>
  transformation_log: string[]
  retry_count: number
}

export type ReconciliationCheck = {
  table_a: string
  column_a: string
  table_b: string
  column_b: string
  orphan_rate_before: number
  orphan_rate_after: number
  within_tolerance: boolean
}

export type CleanRemainingResult = {
  results: TableCleaningStatus[]
  reconciliation: ReconciliationCheck[]
}

export type IntegrityResult = {
  passed: boolean
  report: Record<string, any>
}

export type SaveResult = {
  saved: boolean
  schema_name: string
  tables: string[]
}

export function useSchemaDiscovery(sessionId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: () => call<SchemaDiscoveryResult>(`/api/v1/pipeline/${sessionId}/schema-discovery`, { method: 'POST' }),
  })
}

export function useSubmitRelationships(sessionId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (decisions: RelationshipDecision[]) =>
      call<RelationshipOut[]>(`/api/v1/pipeline/${sessionId}/relationships`, { method: 'POST', json: { decisions } }),
  })
}

export function usePlanTable(sessionId: string, tableName: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: () => call<PlanResult>(`/api/v1/pipeline/${sessionId}/tables/${tableName}/plan`, { method: 'POST', json: {} }),
  })
}

export function useEditPlan(sessionId: string, tableName: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (cleaning_plan: CleaningPlanStep[]) =>
      call<PlanResult>(`/api/v1/pipeline/${sessionId}/tables/${tableName}/plan`, { method: 'PATCH', json: { cleaning_plan } }),
  })
}

export function usePlanStepOpinion(sessionId: string, tableName: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (step_description: string) =>
      call<{ opinion: string }>(`/api/v1/pipeline/${sessionId}/tables/${tableName}/plan/opinion`, {
        method: 'POST', json: { step_description },
      }),
  })
}

export function useCleanTable(sessionId: string, tableName: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: () => call<CleanResult>(`/api/v1/pipeline/${sessionId}/tables/${tableName}/clean`, { method: 'POST', json: {} }),
  })
}

export function useCleanRemaining(sessionId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: () => call<CleanRemainingResult>(`/api/v1/pipeline/${sessionId}/clean-remaining`, { method: 'POST' }),
  })
}

export function useReconciliation(sessionId: string, enabled: boolean) {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['reconciliation', sessionId],
    queryFn: () => call<ReconciliationCheck[]>(`/api/v1/pipeline/${sessionId}/reconciliation`),
    enabled: ready && enabled,
  })
}

export function useIntegrity(sessionId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: () => call<IntegrityResult>(`/api/v1/pipeline/${sessionId}/integrity`, { method: 'POST' }),
  })
}

export function useSaveToProject(sessionId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: () => call<SaveResult>(`/api/v1/pipeline/${sessionId}/save`, { method: 'POST' }),
  })
}

File: backend/app/schemas/pipeline.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class TablePreview(BaseModel):
    name: str
    rows: int
    columns: list[str]
    preview: list[dict[str, Any]]


class IngestResponse(BaseModel):
    pipeline_session_id: str
    project_id: str
    primary_table: str
    tables: list[TablePreview]


class GSheetIngestRequest(BaseModel):
    service_account_json: str = Field(description="Raw JSON contents of the GCP service account key")
    sheet_url_or_id: str


class DbLinkTestRequest(BaseModel):
    dialect: str = Field(pattern="^(postgresql|mysql|mssql)$")
    host: str
    port: int
    database: str
    username: str
    password: str


class DbLinkIngestRequest(DbLinkTestRequest):
    tables: list[str] | None = None
    save_connection: bool = False


class TableProfileOut(BaseModel):
    name: str
    row_count: int
    column_count: int
    columns: list[str]


class RelationshipCandidateOut(BaseModel):
    table_a: str
    column_a: str
    table_b: str
    column_b: str
    confidence: float
    evidence: dict[str, float]
    source: str
    reasoning: str
    validation: dict[str, Any]


class SchemaDiscoveryResponse(BaseModel):
    tables: list[TableProfileOut]
    auto: list[RelationshipCandidateOut]
    review: list[RelationshipCandidateOut]
    manual: list[RelationshipCandidateOut]


class RelationshipDecision(BaseModel):
    table_a: str
    column_a: str
    table_b: str
    column_b: str
    confidence: float = 0.0
    evidence: dict[str, float] = Field(default_factory=dict)
    status: str = Field(pattern="^(approved|rejected)$")


class RelationshipsRequest(BaseModel):
    decisions: list[RelationshipDecision]


class RelationshipOut(BaseModel):
    table_a: str
    column_a: str
    table_b: str
    column_b: str
    confidence: float
    status: str
    approved_by: str


class PlanRequest(BaseModel):
    planner_model: str | None = None
    coder_model: str | None = None


class CleaningPlanStep(BaseModel):
    id: str
    description: str
    source: str = Field(default="generated", pattern="^(generated|manual|detector|fallback)$")
    # A step is normally a typed operator from tools.cleaning_ops.REGISTRY.
    # An empty `op` marks a free-text step, which falls back to generated code.
    op: str = ""
    columns: list[str] = Field(default_factory=list)
    params: dict[str, Any] = Field(default_factory=dict)


class PlanResponse(BaseModel):
    table_name: str
    profile: dict[str, Any]
    cleaning_plan: list[CleaningPlanStep]


class PlanEditRequest(BaseModel):
    cleaning_plan: list[CleaningPlanStep]


class PlanStepOpinionRequest(BaseModel):
    step_description: str
    planner_model: str | None = None


class PlanStepOpinionResponse(BaseModel):
    opinion: str


class CleanRequest(BaseModel):
    planner_model: str | None = None
    coder_model: str | None = None


class CleanResponse(BaseModel):
    table_name: str
    generated_code: str
    success: bool
    error: str | None
    validation_report: dict[str, Any]
    transformation_log: list[str]
    retry_count: int
    preview: TablePreview | None
    # Measured per-step audit trail (counts computed by diffing the frame, not
    # reported by the code that changed it). See models/cleaning_ops.py.
    ledger: dict[str, Any] | None = None


class TableCleaningStatus(BaseModel):
    table_name: str
    success: bool
    error: str | None
    validation_report: dict[str, Any]
    transformation_log: list[str]
    retry_count: int
    ledger: dict[str, Any] | None = None


class ReconciliationCheckOut(BaseModel):
    table_a: str
    column_a: str
    table_b: str
    column_b: str
    orphan_rate_before: float
    orphan_rate_after: float
    within_tolerance: bool


class CleanRemainingResponse(BaseModel):
    results: list[TableCleaningStatus]
    reconciliation: list[ReconciliationCheckOut]


class IntegrityResponse(BaseModel):
    passed: bool
    report: dict[str, Any]


class SaveResponse(BaseModel):
    saved: bool
    schema_name: str
    tables: list[str]
